src/libraries/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs

```diff
@@ -50,6 +50,7 @@ internal void SetUnmanagedStructures(object? objectsToPin)
 
             unsafe
             {
+                Debug.Assert(OperatingSystem.IsWindows());
```

Comment:
It's strange to me that we're having to do such asserts, given that this whole file is ".Windows.cs" and is only included in Windows builds.  Is this the best answer?

Reply:
We should automatically mark the Windows-specific implementation assemblies with `[assembly: SupportedOSPlatform("windows")]`.

Reply:
So we need to setup infrastructure to mark the Windows specific builds as Windows and the Unix specific builds as "relevant OS"?

Reply:
Do we have a way of identifying assemblies that are truly "any platform" vs one that must build for Unix vs Windows or Linux vs MacOS vs Windows, etc?

Reply:
> We should automatically mark the Windows-specific implementation assemblies with `[assembly: SupportedOSPlatform("windows")]`.

We tried heading down that path in #41209 but it was snowballing and we identified that it could inadvertently affect the ref assemblies when we generate those.  I'm open to us taking another crack at it though.

Reply:
> I'm open to us taking another crack at it though.

@jeffhandley right, my assumption was that this issue isn't for .NET 5 anymore which opens up the opportunity to do The Right Thing™

Reply:
> @jeffhandley right, my assumption was that this issue isn't for .NET 5 anymore which opens up the opportunity to do The Right Thing™

Sounds good -- let's proceed with this approach for now and do The Right Thing™️ later.  That will let us have the analyzer on right away.

Reply:
I filed https://github.com/dotnet/runtime/issues/41901 to track the future investment.